### PR TITLE
Replacing thread_local in qgscoordinatetransform_p 

### PR DESCRIPTION
**Pull request by t-hey**
## Description
When using thread_local in qgscoordinatetransform_p, it crashes QGIS in a very short time when a render thread is destroyed, while in debugging mode in Qt Creator, this makes debugging very difficult . When  running QGIS in non debugging mode from Qt Creator it crash, at the same place in code, when closing QGIS.
It crashes in qgscoordinatetransform_p.cpp line 41 when the QgsProjContextStore's destructor is called 
and the the  QgsProjContextStore instance could not be created in the first place.

When using QThreadStorage everything works as expected.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nyalldawson** [comment]
Thanks @t-hey!

For reference, this PR is designed to fix a crash encountered when building QGIS under mingw, caused by:

https://sourceforge.net/p/mingw-w64/bugs/445/
https://sourceforge.net/p/mingw-w64/bugs/527/
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80816

Note that this relates also to https://issues.qgis.org/issues/17351 (build failure on BSD)

2.18 uses QThreadStorage to handle this and I'm unaware of any issues in 2.18 which relate.

As far as I'm aware, using thread_local in master is working correctly on all our officially supported build configurations.

The open question is whether we replace the std thread_local with QThreadStorage on all platforms, or whether we ifdef this in for mingw/OpenBSD builds. 

My preference would lean to the latter. But my preference is driven by a growing distrust in Qt upstream to keep any class which duplicates std functionality well maintained and not to drop them in future releases.

**Comment by t-hey** [comment]
@nyalldawson I see the Travis build failed :-(
 I looked at the code again and line 208 and 209 is suspect. 
Should there be two inserts into mProjProjection? 
My feeling is that 
`mProjProjections.insert( reinterpret_cast< uintptr_t>( mProjContext.localData()->get() ), res );`
should be removed.

File: src/core/qgscoordinatetransform_p.cpp [hunk 1]
```diff
@@ -196,9 +203,10 @@ QPair<projPJ, projPJ> QgsCoordinateTransformPrivate::threadLocalProjData()
   // proj projections don't exist yet, so we need to create
   mProjLock.unlock();
   mProjLock.lockForWrite();
-  QPair<projPJ, projPJ> res = qMakePair( pj_init_plus_ctx( mProjContext.get(), mSourceProjString.toUtf8() ),
-                                         pj_init_plus_ctx( mProjContext.get(), mDestProjString.toUtf8() ) );
-  mProjProjections.insert( reinterpret_cast< uintptr_t>( mProjContext.get() ), res );
+  QPair<projPJ, projPJ> res = qMakePair( pj_init_plus_ctx( pContext, mSourceProjString.toUtf8() ),
+                                             pj_init_plus_ctx( pContext, mDestProjString.toUtf8() ) );
+  mProjProjections.insert( reinterpret_cast< uintptr_t>( pContext ), res );
+  mProjProjections.insert( reinterpret_cast< uintptr_t>( mProjContext.localData()->get() ), res );
```
Comment (nyalldawson) [inline src/core/qgscoordinatetransform_p.cpp]: Yes, this line should be removed. Leave the pContext one.

**Comment by nyalldawson** [comment]
@t-hey there is a duplicate line. But Travis is ultimately failing the code formatting test.

This can be fixed by enabled WITH_ASTYLE in your build, and then running the pre-commit script (see https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#proper-formatting )

**Comment by landryb** [comment]
Testing this PR on OpenBSD to see if it fixes https://issues.qgis.org/issues/17351

**Comment by landryb** [comment]
Builds fine on OpenBSD when added on top of master, and qgis2.99 starts fine with it.

**Comment by lbartoletti** [comment]
Hi,
I will also try it for FreeBSD this week.

**Comment by nyalldawson** [comment]
Superseded by #5541

**Pull request closed without merging**